<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cosmosR">
<title>COSMOS-tutorial • cosmosR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="COSMOS-tutorial">
<meta property="og:description" content="cosmosR">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-PV71G2Q3SF"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-PV71G2Q3SF');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cosmosR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tutorial.html">COSMOS-tutorial</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/saezlab/COSMOSR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/saezlab">
    <span class="fab fa fab fa-twitter fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://saezlab.org">
    <span class="fas fa fas fa-university"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="tutorial_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="tutorial_files/vis-9.1.0/vis-network.min.css" rel="stylesheet">
<script src="tutorial_files/vis-9.1.0/vis-network.min.js"></script><script src="tutorial_files/visNetwork-binding-2.1.0/visNetwork.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>COSMOS-tutorial</h1>
                        <h4 data-toc-skip class="author">A. Dugourd, A.
Gabor and K. Zirngibl</h4>
            
            <h4 data-toc-skip class="date">11/10/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/saezlab/COSMOSR/blob/HEAD/vignettes/tutorial.Rmd" class="external-link"><code>vignettes/tutorial.Rmd</code></a></small>
      <div class="d-none name"><code>tutorial.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation-and-dependency">Installation and dependency<a class="anchor" aria-label="anchor" href="#installation-and-dependency"></a>
</h2>
<p>COSMOS is dependent on CARNIVAL for exhibiting the signalling pathway
optimisation. CARNIVAL requires the interactive version of IBM Cplex or
CBC-COIN solver as the network optimiser. The IBM ILOG Cplex is freely
available through Academic Initiative <a href="https://www.ibm.com/products/ilog-cplex-optimization-studio" class="external-link">here</a>.
The <a href="https://projects.coin-or.org/Cbc" class="external-link">CBC</a> solver is open
source and freely available for any user, but has a significantly lower
performance than CPLEX. Obtain CBC executable directly usable for cosmos
<a href="https://ampl.com/products/solvers/open-source/#cbc" class="external-link">here</a>.
Alternatively for small networks, users can rely on the freely available
<a href="https://cran.r-project.org/web/packages/lpSolve/index.html" class="external-link">lpSolve
R-package</a>, which is automatically installed with the package.</p>
<p>In this tutorial we use lpSolve, but we strongly recommend to obtain
a license for CPLEX.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install from bioconductor</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"cosmosR"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># install the newest (development) version from GitHub</span></span>
<span><span class="co"># install.packages("remotes")</span></span>
<span><span class="co"># install CARNIVAL from github </span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"saezlab/CARNIVAL"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"saezlab/cosmosR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>COSMOS (Causal Oriented Search of Multi-Omic Space) is a method that
integrates phosphoproteomics, transcriptomics, and metabolomics data
sets. COSMOS leverages extensive prior knowledge of signaling pathways,
metabolic networks, and gene regulation with computational methods to
estimate activities of transcription factors and kinases as well as
network-level causal reasoning. This pipeline can provide mechanistic
explanations for experimental observations across multiple omic data
sets.</p>
<div class="figure">
<img src="../reference/figures/intro_data.png" alt=""><p class="caption">data_intro_figure</p>
</div>
<p>Essentially, COSMOS has 3 main components:</p>
<ul>
<li><p>The prior knowledge: The goal of COSMOS is essentially to connect
various deregulation events together with mechanistic hypotheses. Those
mechanistic hypotheses are basically known potential interactions
between molecular biology features, such as proteins and metabolites. An
example of such interaction can be the activation of MTOR by AKT, in the
EGFR canonical pathway. Thousands of such potential interaction can be
assembled together to form a large network of interactions. Such
networks are thus called prior knowledge networks (PKN), because they
summarize large amounts of prior knowledge in the form of a network. In
the context of COSMOS this interactions come from the Omnipath database,
STICHdb and the recon3D reaction network. With respect to the example
interaction between MTOR and AKT, the question that is usually asked
when presented with a given experimental context is: while MTOR can
activate AKT in general, is this interaction relevant in my experimental
context? This brings us to the second component of COSMOS.</p></li>
<li><p>The data: Information about deregulated transcription factors
(TF), kinases/phosphophatases and metabolites can be extracted from
transcriptomic, phospho-proteomic or metabolomic data respectively,
using footprint analysis methods (see:<a href="https://github.com/saezlab/transcriptutorial" class="external-link uri">https://github.com/saezlab/transcriptutorial</a> (Especially
chapter 4)). Such information can be quite useful by itself, to
characterize which TFs or kinases are specifically deregulated in cancer
cells for example. the goal of COSMOS is to find how these different
deregulation events can be connected together. Going back to the example
of MTOR and AKT, if I know that AKT can potentially activate MTOR, and I
observe in my data that both AKT and MTOR are significantly activated in
cancer tissue compared to their healthy counterpart, there is a
reasonable chance that this interaction is active in the cancer tissue
I’m studying. The problem is that data often contain information about
hundreds of up and down-regulated activities at the same time, and
thousands of potential interactions can potentially explain them. In
order to find the optimal set of interactions that are relevant to
explain the deregulation observed in my data, an algorithm is
needed.</p></li>
<li><p>The algorithm: The problem described in the data part is
formulated as an integer linear optimisation problem. The PKN and the
data can be abstracted as a large set of integer variables operating
together in an even larger set of equations (see <a href="https://www.nature.com/articles/s41540-019-0118-z" class="external-link uri">https://www.nature.com/articles/s41540-019-0118-z</a> for
more info). Given a set of fixed value for the variable (that is, a
subset of the known deregulation), we need to find the values for the
other sets of variable (corresponding to proteins and metabolites for
which we have no information in our data) that can lead to a satisfying
result. A satisfying results in the context of COSMOS represent a
sub-network of interactions that is as small as possible while
coherently explaining as many of our known deregulation as possible.
Thus, the set of equation mentioned earlier needs to be solved, using an
integer linear programming solver. The solvers that are currently
available for COSMOS are lpsolve, cbc and cplex. In general, CPLEX is
preferred over CBC for most real case applications, while lpsolve is
mainly used for testing with small networks and sets of
measurements.</p></li>
</ul>
<div class="figure">
<img src="../reference/figures/summary.png" alt=""><p class="caption">Summary figure</p>
</div>
</div>
<div class="section level2">
<h2 id="tutorial-section-signaling-to-metabolism">Tutorial section: signaling to metabolism<a class="anchor" aria-label="anchor" href="#tutorial-section-signaling-to-metabolism"></a>
</h2>
<p>First, we load the package</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/saezlab/COSMOSR" class="external-link">cosmosR</a></span><span class="op">)</span></span></code></pre></div>
<p>In this part, we can set up the options for the CARNIVAL run, such as
timelimit and min gap tolerance.</p>
<p>The user should provide a path to its CPLEX/cbc executable.</p>
<p>You can check the CARNIVAL_options variable to see all possible
options that can be adjusted</p>
<p>In this example, we will use the built-in solver lpSolve. User should
be aware that lpSolve should ONLY be used for TESTS. To obtain
meaningful results, best solver is cplex, or cbc if not possible.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CARNIVAL_options</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/default_CARNIVAL_options.html">default_CARNIVAL_options</a></span><span class="op">(</span>solver <span class="op">=</span> <span class="st">"lpSolve"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># To use CBC</span></span>
<span><span class="co"># CARNIVAL_options &lt;- default_CARNIVAL_options(solver = "cbc")</span></span>
<span><span class="co"># CARNIVAL_options$solverPath &lt;- "~/Documents/cbc"</span></span>
<span><span class="co"># CARNIVAL_options$threads &lt;- 2</span></span>
<span><span class="co"># CARNIVAL_options$mipGAP &lt;- 0.05</span></span>
<span></span>
<span><span class="co"># To use CPLEX:</span></span>
<span><span class="co"># CARNIVAL_options &lt;- default_CARNIVAL_options(solver = "cplex")</span></span>
<span><span class="co"># CARNIVAL_options$solverPath &lt;- "C:/Program Files/CPLEX_solver/cplex/bin/x64_win64/cplex.exe"</span></span>
<span><span class="co"># CARNIVAL_options$threads &lt;- 2</span></span>
<span><span class="co"># CARNIVAL_options$mipGAP &lt;- 0.05</span></span>
<span></span>
<span></span>
<span><span class="va">CARNIVAL_options</span><span class="op">$</span><span class="va">timelimit</span> <span class="op">&lt;-</span> <span class="fl">3600</span></span></code></pre></div>
<p>In the next section, we prepare the input to run cosmosR. The
signaling inputs are the result of footprint based TF and kinase
activity estimation. For more info on TF activity estimation from
transcriptomic data, see:<a href="https://github.com/saezlab/transcriptutorial" class="external-link uri">https://github.com/saezlab/transcriptutorial</a> (Especially
chapter 4)</p>
<p>Here we use of toy PKN, to see the full meta PKN, you can load it
with data(meta_network). To see how is the meta_PKN assembled, see: <a href="https://github.com/saezlab/meta_PKN_BIGG.git" class="external-link uri">https://github.com/saezlab/meta_PKN_BIGG.git</a></p>
<p>The metabolites in the prior knowledge network are identified as
XMetab__HMDBid_compartment or XMetab__BIGGid_compartment (for example
“Metab__HMDB0000190_c”). The compartment code is the BIGG model standard
(r, c, e, x, m, l, n, g). Thus we will first need to map whatever
identifier for metabolite the data has to the one of the network. Genes
are identified as gene symboles (in the signaling part of network) or
Gene####__symbole (in the reaction network part of network).</p>
<p>The maximum network depth will define the maximum number of step
downstream of kinase/TF COSMOS will look for deregulated metabolites.
Good first guess for max depth could be around between 4 and 6 (here it
is 15 for the toy dataset)</p>
<p>The differential experession data is used to filter out wrong
TF-target interactions in this context after a pre-optimisation.</p>
<p>The list of genes in the differential expression data will also be
used as a reference to define which genes are expressed or not (all
genes in the diff_expression_data are considered expressed, and genes
that are no in diff_expression_data are removed from the network).</p>
<p>Here, the CARNIVAL_options$timelimit is set for the pre-optimisation.
Indeed, if the “filter_tf_gene_interaction_by_optimization” parameter of
the “preprocess_COSMOS_signaling_to_metabolism” function is set to TRUE,
COSMOS can perform a first optimisation run in order to generate a
preliminary solution network that can be used for filtering out
incoherences between TF activities and downstream target
expressions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">toy_network</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">toy_signaling_input</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">toy_metabolic_input</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">toy_RNA</span><span class="op">)</span></span>
<span><span class="va">test_for</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_COSMOS_signaling_to_metabolism.html">preprocess_COSMOS_signaling_to_metabolism</a></span><span class="op">(</span>meta_network <span class="op">=</span> <span class="va">toy_network</span>,</span>
<span>                                        signaling_data <span class="op">=</span> <span class="va">toy_signaling_input</span>,</span>
<span>                                        metabolic_data <span class="op">=</span> <span class="va">toy_metabolic_input</span>,</span>
<span>                                                      diff_expression_data <span class="op">=</span> <span class="va">toy_RNA</span>,</span>
<span>                                                      maximum_network_depth <span class="op">=</span> <span class="fl">15</span>,</span>
<span>                                                      remove_unexpressed_nodes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                      CARNIVAL_options <span class="op">=</span> <span class="va">CARNIVAL_options</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "COSMOS: all 3 signaling nodes from data were found in the meta PKN"</span></span>
<span><span class="co">## [1] "COSMOS: all 2 metabolic nodes from data were found in the meta PKN"</span></span>
<span><span class="co">## [1] "COSMOS: 2975 of the 9300 genes in expression data were found as transcription factor target"</span></span>
<span><span class="co">## [1] "COSMOS: 2975 of the 5321 transcription factor targets were found in expression data"</span></span>
<span><span class="co">## [1] "COSMOS: removing unexpressed nodes from PKN..."</span></span>
<span><span class="co">## [1] "COSMOS: 0 interactions removed"</span></span>
<span><span class="co">## [1] "COSMOS: removing nodes that are not reachable from inputs within 15 steps"</span></span>
<span><span class="co">## [1] "COSMOS: 0 from  101 interactions are removed from the PKN"</span></span>
<span><span class="co">## [1] "COSMOS: removing nodes that are not observable by measurements within 15 steps"</span></span>
<span><span class="co">## [1] "COSMOS: 52 from  101 interactions are removed from the PKN"</span></span>
<span><span class="co">## [1] "COSMOS: 2 input/measured nodes are not in PKN any more: USF1, SRF and 0 more."</span></span>
<span><span class="co">## [1] "COSMOS:  0 interactions are removed from the PKN based on consistency check between TF activity and gene expression"</span></span>
<span><span class="co">## [1] "lpSolve does not scale well with large PKNs. This solver is mainly for testing purposes. To run COSMSO, we recommend using cplex, or cbc solvers."</span></span>
<span><span class="co">## [1] "lpSolve does not scale well with large PKNs. This solver is mainly for testing purposes. To run COSMSO, we recommend using cplex, or cbc solvers."</span></span>
<span><span class="co">## [1] "COSMOS:  0 interactions are removed from the PKN based on consistency check between TF activity and gene expression"</span></span>
<span><span class="co">## [1] "COSMOS: all 1 signaling nodes from data were found in the meta PKN"</span></span>
<span><span class="co">## [1] "COSMOS: all 2 metabolic nodes from data were found in the meta PKN"</span></span>
<span><span class="co">## [1] "COSMOS: 2975 of the 9300 genes in expression data were found as transcription factor target"</span></span>
<span><span class="co">## [1] "COSMOS: 2975 of the 5321 transcription factor targets were found in expression data"</span></span></code></pre>
<p>In this part, we can set up the options for the actual run, such as
timelimit and min gap tolerance.</p>
<p>The running time should be much higher here than in pre-optimisation.
You can increase the number of threads to use if you have many available
CPUs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CARNIVAL_options</span><span class="op">$</span><span class="va">timelimit</span> <span class="op">&lt;-</span> <span class="fl">14400</span></span>
<span><span class="va">CARNIVAL_options</span><span class="op">$</span><span class="va">mipGAP</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">CARNIVAL_options</span><span class="op">$</span><span class="va">threads</span> <span class="op">&lt;-</span> <span class="fl">2</span></span></code></pre></div>
<p>This is where cosmosR run.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_result_for</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_COSMOS_signaling_to_metabolism.html">run_COSMOS_signaling_to_metabolism</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">test_for</span>,</span>
<span>                                                      CARNIVAL_options <span class="op">=</span> <span class="va">CARNIVAL_options</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "lpSolve does not scale well with large PKNs. This solver is mainly for testing purposes. To run COSMSO, we recommend using cplex, or cbc solvers."</span></span>
<span><span class="co">## [1] "lpSolve does not scale well with large PKNs. This solver is mainly for testing purposes. To run COSMSO, we recommend using cplex, or cbc solvers."</span></span></code></pre>
<p>Finally, we process the results of the first cosmosR run, to
translate gene names and metabolites name.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">formated_result_for</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_COSMOS_res.html">format_COSMOS_res</a></span><span class="op">(</span><span class="va">test_result_for</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorial-section-metabolism-to-signaling">Tutorial section: metabolism to signaling<a class="anchor" aria-label="anchor" href="#tutorial-section-metabolism-to-signaling"></a>
</h2>
<p>Before we run the metabolism to signaling part, we need to prepare
again the inputs.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CARNIVAL_options</span><span class="op">$</span><span class="va">timelimit</span> <span class="op">&lt;-</span> <span class="fl">3600</span></span>
<span><span class="va">CARNIVAL_options</span><span class="op">$</span><span class="va">mipGAP</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">CARNIVAL_options</span><span class="op">$</span><span class="va">threads</span> <span class="op">&lt;-</span> <span class="fl">2</span></span></code></pre></div>
<p>Now that the correct time is set up for the pre-optimisation run, we
can prepare the inputs.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_back</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_COSMOS_metabolism_to_signaling.html">preprocess_COSMOS_metabolism_to_signaling</a></span><span class="op">(</span>meta_network <span class="op">=</span> <span class="va">toy_network</span>,</span>
<span>                                        signaling_data <span class="op">=</span> <span class="va">toy_signaling_input</span>,</span>
<span>                                        metabolic_data <span class="op">=</span> <span class="va">toy_metabolic_input</span>,</span>
<span>                                                       diff_expression_data <span class="op">=</span> <span class="va">toy_RNA</span>,</span>
<span>                                                       maximum_network_depth <span class="op">=</span> <span class="fl">15</span>,</span>
<span>                                                       remove_unexpressed_nodes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                                       CARNIVAL_options <span class="op">=</span> <span class="va">CARNIVAL_options</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "COSMOS: all 3 signaling nodes from data were found in the meta PKN"</span></span>
<span><span class="co">## [1] "COSMOS: all 2 metabolic nodes from data were found in the meta PKN"</span></span>
<span><span class="co">## [1] "COSMOS: 2975 of the 9300 genes in expression data were found as transcription factor target"</span></span>
<span><span class="co">## [1] "COSMOS: 2975 of the 5321 transcription factor targets were found in expression data"</span></span>
<span><span class="co">## [1] "COSMOS: removing nodes that are not reachable from inputs within 15 steps"</span></span>
<span><span class="co">## [1] "COSMOS: 0 from  101 interactions are removed from the PKN"</span></span>
<span><span class="co">## [1] "COSMOS: removing nodes that are not observable by measurements within 15 steps"</span></span>
<span><span class="co">## [1] "COSMOS: 54 from  101 interactions are removed from the PKN"</span></span>
<span><span class="co">## [1] "COSMOS: 1 input/measured nodes are not in PKN any more: Metab__HMDB0000190_c and 0 more."</span></span>
<span><span class="co">## [1] "COSMOS:  0 interactions are removed from the PKN based on consistency check between TF activity and gene expression"</span></span>
<span><span class="co">## [1] "lpSolve does not scale well with large PKNs. This solver is mainly for testing purposes. To run COSMSO, we recommend using cplex, or cbc solvers."</span></span>
<span><span class="co">## [1] "lpSolve does not scale well with large PKNs. This solver is mainly for testing purposes. To run COSMSO, we recommend using cplex, or cbc solvers."</span></span>
<span><span class="co">## [1] "COSMOS:  0 interactions are removed from the PKN based on consistency check between TF activity and gene expression"</span></span>
<span><span class="co">## [1] "COSMOS: all 3 signaling nodes from data were found in the meta PKN"</span></span>
<span><span class="co">## [1] "COSMOS: all 1 metabolic nodes from data were found in the meta PKN"</span></span>
<span><span class="co">## [1] "COSMOS: 2975 of the 9300 genes in expression data were found as transcription factor target"</span></span>
<span><span class="co">## [1] "COSMOS: 2975 of the 5321 transcription factor targets were found in expression data"</span></span></code></pre>
<p>Then we can run cosmosR to connect metabolism to signaling. The
running time here usually needs to be longer, as this problem seems to
be harder to solve for CPLEX.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CARNIVAL_options</span><span class="op">$</span><span class="va">timelimit</span> <span class="op">&lt;-</span> <span class="fl">28800</span></span>
<span></span>
<span><span class="va">test_result_back</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_COSMOS_metabolism_to_signaling.html">run_COSMOS_metabolism_to_signaling</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">test_back</span>,</span>
<span>                                                       CARNIVAL_options <span class="op">=</span> <span class="va">CARNIVAL_options</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "lpSolve does not scale well with large PKNs. This solver is mainly for testing purposes. To run COSMSO, we recommend using cplex, or cbc solvers."</span></span>
<span><span class="co">## [1] "lpSolve does not scale well with large PKNs. This solver is mainly for testing purposes. To run COSMSO, we recommend using cplex, or cbc solvers."</span></span></code></pre>
<p>Finally we can format the result of the backward run as well (same as
for forward run)</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">formated_result_back</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_COSMOS_res.html">format_COSMOS_res</a></span><span class="op">(</span><span class="va">test_result_back</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorial-section-merge-forward-and-backward-networks-and-visualise-network">Tutorial section: Merge forward and backward networks and visualise
network<a class="anchor" aria-label="anchor" href="#tutorial-section-merge-forward-and-backward-networks-and-visualise-network"></a>
</h2>
<p>Here we simply take the union of forward and backward runs to create
a full network solution lopping between signaling, gene-regulation and
metabolism. Since there is an overlap between the result network of
forward and backward run, you may optionally want to check if there are
any node sign that are incoherent in the overlap between the two
solutions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_sif</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">formated_result_for</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">formated_result_back</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">full_sif</span> <span class="op">&lt;-</span> <span class="va">full_sif</span><span class="op">[</span><span class="va">full_sif</span><span class="op">$</span><span class="va">Weight</span><span class="op">&gt;</span><span class="fl">0</span>,<span class="op">]</span></span>
<span><span class="va">full_attributes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">formated_result_for</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">formated_result_back</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">full_sif</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">full_sif</span><span class="op">)</span></span>
<span><span class="va">full_attributes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">full_attributes</span><span class="op">)</span></span></code></pre></div>
<p>This function will generate a dynamic network plot centered on a
given node of the network solution, and connecting it to measured nodes
in the given range (here 7 steps).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">network_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/display_node_neighboorhood.html">display_node_neighboorhood</a></span><span class="op">(</span>central_node <span class="op">=</span> <span class="st">'Metab__D-Glucitol_c'</span>, </span>
<span>                                           sif <span class="op">=</span> <span class="va">full_sif</span>, </span>
<span>                                           att <span class="op">=</span> <span class="va">full_attributes</span>, </span>
<span>                                           n <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span></span>
<span><span class="va">network_plot</span></span></code></pre></div>
<div id="htmlwidget-29f13a6c050a401edb91" style="width:1600px;height:1600px;" class="visNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-29f13a6c050a401edb91">{"x":{"nodes":{"id":["Enzyme1338__AKR1A1","Enzyme7338__LDHB_LDHA_reverse","Enzyme9049__SLC2A1_reverse","FGFR1","LDHA","MAPK1","MAPK14","Metab__D-Glucitol_c","Metab__D-Glucose_c","Metab__L-Lactic acid_c","MYC","SLC2A1","SRF","USF1"],"NodeType":["","","","","","","","M","","M","P","","M","M"],"ZeroAct":[0,0,0,0,0,0,0,0,0,0,0,0,0,0],"UpAct":[1,1,1,1,1,0,1,1,1,1,1,1,1,1],"DownAct":[0,0,0,0,0,1,0,0,0,0,0,0,0,0],"AvgAct":[1,1,1,1,1,-1,1,1,1,1,1,1,1,1],"measured":[0,0,0,0,0,0,0,1,0,1,1,0,1,1],"Activity":[1,1,1,1,1,-1,1,1,1,1,1,1,1,1],"label":["Enzyme1338__AKR1A1","Enzyme7338__LDHB_LDHA_reverse","Enzyme9049__SLC2A1_reverse","FGFR1","LDHA","MAPK1","MAPK14","Metab__D-Glucitol_c","Metab__D-Glucose_c","Metab__L-Lactic acid_c","MYC","SLC2A1","SRF","USF1"],"color":["green","green","green","green","green","red","green","green","green","green","green","green","green","green"],"shape":["dot","dot","dot","dot","dot","dot","dot","dot","dot","dot","dot","dot","dot","dot"],"shadow":[false,false,false,false,false,false,false,true,false,true,true,false,true,true]},"edges":{"from":["Enzyme1338__AKR1A1","Enzyme9049__SLC2A1_reverse","Metab__D-Glucose_c","MYC","SLC2A1","FGFR1","Enzyme7338__LDHB_LDHA_reverse","LDHA","MAPK1","MAPK14","MAPK14","Metab__D-Glucitol_c","MAPK14","MAPK14"],"to":["Metab__D-Glucitol_c","Metab__D-Glucose_c","Enzyme1338__AKR1A1","SLC2A1","Enzyme9049__SLC2A1_reverse","LDHA","Metab__L-Lactic acid_c","Enzyme7338__LDHB_LDHA_reverse","FGFR1","MAPK1","MYC","MAPK14","SRF","USF1"],"sign":[1,1,1,1,1,1,1,1,-1,-1,1,1,1,1],"weigth":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"color":["grey","grey","grey","grey","grey","grey","grey","grey","grey","grey","grey","grey","grey","grey"],"arrows.to.type":["arrow","arrow","arrow","arrow","arrow","arrow","arrow","arrow","circle","circle","arrow","arrow","arrow","arrow"],"enabled":[true,true,true,true,true,true,true,true,true,true,true,true,true,true],"scaleFactor":[1,1,1,1,1,1,1,1,1,1,1,1,1,1]},"nodesToDataframe":true,"edgesToDataframe":true,"options":{"width":"100%","height":"100%","nodes":{"shape":"dot"},"manipulation":{"enabled":false}},"groups":null,"width":1600,"height":1600,"idselection":{"enabled":true,"style":"width: 200px; height: 26px;\n                                              background: #f8f8f8;\n                                              color: darkblue;\n                                              border:none;\n                                              outline:none;","useLabels":true,"main":"Select by id"},"byselection":{"enabled":false,"style":"width: 150px; height: 26px","multiple":false,"hideColor":"rgba(200,200,200,0.5)","highlight":false},"main":null,"submain":null,"footer":null,"background":"rgba(0, 0, 0, 0)","highlight":{"enabled":true,"hoverNearest":false,"degree":1,"algorithm":"all","hideColor":"rgba(200,200,200,0.5)","labelOnly":true},"collapse":{"enabled":false,"fit":false,"resetHighlight":true,"clusterOptions":null,"keepCoord":true,"labelSuffix":"(cluster)"}},"evals":[],"jsHooks":[]}</script><p>Here is how this network can be intepreted (this is purely
illustrative, as some of those interaction may be incorectly signed
because lpsolve can only use positive interactions):</p>
<p>This network represents the flow of activities that can connect MYC
up-regulation with Glucitol (Sorbitol) accumulation. Here, NFKB1 can
upregulate the expression of SLC2A1, which in turn transport more
glucose in the cytoplasm. The increase transport of glucose can lead to
more glucose being avlaible for conversion into glucitol by the AKR1A
enzyme. Interestingly, glucitol is a now activator of MAPK14, thus
leading to the appearance of a positive feedback loop connecting MYC,
glucitol and MAPK14.</p>
<p>It is important to understand that each of this links is
hypothetical. The come from a larger pool of potential molecular
interactions present in multiple online databases and compiled in
omnipath, STITCH and recon metabolic network. They exist in the
literature and are interactions that are known to potentially exists in
other experimental contexts. Thus, COSMOS compile all those potential
interactions together and proposes a coherent set that can explain the
data at hand.</p>
<p>Those links should however be considered only as potential
mechanistic connections, and will need to be further confirmed
experimentally. Those interactions can be searched in the literature to
see in which other disease or experimental context they have been shown
to be relevant. Taken together, multiple interactions can help to build
a biological story that can guide further underatanding of the
underlying biology and decide on future experiments.</p>
</div>
<div class="section level2">
<h2 id="tutorial-section-over-representation-analysis">Tutorial section: Over Representation Analysis<a class="anchor" aria-label="anchor" href="#tutorial-section-over-representation-analysis"></a>
</h2>
<p>Often it is useful to perform an Over Representation Analysis (ORA)
on the resulting nodes of a COSMOS network as a first analysis step to
get a more functional interpretation on the modeled signaling cascade. A
common way to this is to test whether the selected genes (nodes) in the
COSMOS solution network show statistically significant differences in
comparison to the prior-knowledge network (PKN).</p>
<p>The differentially expressed genes give information about the
cellular processes that are deregulated and if the proportions in
various pathways are SIGNIFICANTLY different from what is expected.In
this way the significant differences between two biological conditions
(e.g. cancer vs. normal tissue, or treatment vs. untreated cells) can be
shown.</p>
<p>Algorithms that perform an ORA are implemented in other R packages
like piano or decoupleR. In addition to a gene set collection these
algorithms require two different lists as inputs: - nodes in the COSMOS
solution network which relate back to the input data
(e.g. transcriptomics, proteomics, metabolomics, fluxomics, or
perturbations) - all nodes (kinases, transcription factors, metabolites)
in the prior-knowledge network (which are used as the background in our
analysis)</p>
<p>In this section we will show how to obtain these two lists from a
formated COSMOS result object.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sif_forward</span> <span class="op">=</span> <span class="va">formated_result_for</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">att_forward</span> <span class="op">=</span> <span class="va">formated_result_for</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">nodes_ORA</span> <span class="op">=</span> <span class="fu"><a href="../reference/extract_nodes_for_ORA.html">extract_nodes_for_ORA</a></span><span class="op">(</span></span>
<span>    sif <span class="op">=</span> <span class="va">sif_forward</span>, </span>
<span>    att <span class="op">=</span> <span class="va">att_forward</span><span class="op">)</span></span></code></pre></div>
<p>Now this forward and backward sets can be used with any ORA
analysis.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] cosmosR_1.3.4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] progress_1.2.2    bcellViper_1.32.0 tidyselect_1.1.2  xfun_0.31        </span></span>
<span><span class="co">##  [5] bslib_0.3.1       lpSolve_5.6.15    purrr_0.3.4       vctrs_0.4.1      </span></span>
<span><span class="co">##  [9] generics_0.1.2    htmltools_0.5.2   yaml_2.3.5        utf8_1.2.2       </span></span>
<span><span class="co">## [13] rlang_1.0.3       pkgdown_2.0.5     jquerylib_0.1.4   pillar_1.7.0     </span></span>
<span><span class="co">## [17] glue_1.6.2        bit64_4.0.5       lifecycle_1.0.1   stringr_1.4.0    </span></span>
<span><span class="co">## [21] ragg_1.2.2        htmlwidgets_1.5.4 visNetwork_2.1.0  memoise_2.0.1    </span></span>
<span><span class="co">## [25] evaluate_0.15     knitr_1.39        CARNIVAL_2.6.0    tzdb_0.3.0       </span></span>
<span><span class="co">## [29] fastmap_1.1.0     dorothea_1.8.0    parallel_4.2.0    fansi_1.0.3      </span></span>
<span><span class="co">## [33] readr_2.1.2       cachem_1.0.6      desc_1.4.1        vroom_1.5.7      </span></span>
<span><span class="co">## [37] jsonlite_1.8.0    systemfonts_1.0.4 fs_1.5.2          bit_4.0.4        </span></span>
<span><span class="co">## [41] textshaping_0.3.6 rjson_0.2.21      hms_1.1.1         digest_0.6.29    </span></span>
<span><span class="co">## [45] stringi_1.7.6     dplyr_1.0.9       rprojroot_2.0.3   cli_3.3.0        </span></span>
<span><span class="co">## [49] tools_4.2.0       magrittr_2.0.3    sass_0.4.1        tibble_3.1.7     </span></span>
<span><span class="co">## [53] crayon_1.5.1      tidyr_1.2.0       pkgconfig_2.0.3   ellipsis_0.3.2   </span></span>
<span><span class="co">## [57] prettyunits_1.1.1 rmarkdown_2.14    R6_2.5.1          igraph_1.3.2     </span></span>
<span><span class="co">## [61] compiler_4.2.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Aurélien Dugourd, Attila Gabor, Katharina Zirngibl.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
